### PR TITLE
Implement ruby-style method visibility

### PR DESCRIPTION
**Pull request by jessedoyle**
This pull request allows method visibility to be specified as follows:

```crystal
class Foo
  def initialize(@arg : Int32)
  end

  private

  def add_one
    @arg + 1
  end

  def add_two
    @arg + 2
  end

  protected

  def something_else; end
end
```

This patch is fully backwards compatible with the previous method visibility syntax:
```crystal
class Foo
  protected

  def a; end          #=> protected method

  private def b; end  #=> private method

  def c; end          #=> protected method
end
```

When the parser encounters a visibility modifier token with a trailing newline, the parser's `@visibility` instance variable is set to the modifier and all following method declarations use the specified visibility.

The `@visibility` variable is reset to `nil` (public) when a `ClassDef` or `ModuleDef` is parsed.

### CODE REVIEW

**Comment by kostya** [comment]
:+1:

**Comment by ozra** [comment]
I like that feature!

**Comment by bcardiff** [comment]
Does this still work with macros? Maybe a spec would be worth it.

**Comment by jessedoyle** [comment]
@bcardiff - Do you mean when instantiating a macro in a private/protected context such as:

```crystal
class Foo
  macro def_name(name)
    def {{name.id}}
      {{name}}
    end
  end

  private

  def_name("bar")

  def something_else; end
end

Foo.new.bar #=> private method 'bar' called for Foo
```

If so, then there are no issues with this pull request. I can add this spec as well.

Forgive me, I don't use macros very often, are there any other cases that I should be checking?

**Comment by bcardiff** [comment]
Yes @jessedoyle, that is the scenario I meant. No need to ask forgiveness. My apologies if I sound terse before. 

It's just a scenario that should be covered / have specs IMO.

I totally like the addition.

**Comment by jessedoyle** [comment]
@bcardiff - Thanks! I just added the extra spec. Just want to make sure that I'm covering all the bases.

**Comment by asterite** [comment]
This feature was left out on purpose. In a long method list it's hard to know what's public and what's private. You have to scroll up to see the private keyword and remember.

However, after using Crystal like this for a while I started missing this feature. I'll have to check with waj.

However, the parser should produce a VisibilityModifier node with a nil expression for this. The parser must produce a node, otherwise when you paste a series of expressions in a macro there will be no trace of this. Then, this logic must be moved to the TypeVisitor. I can guide you if you want to do it, shouldn't be hard.

**Comment by jessedoyle** [comment]
@asterite - Some guidance would be great! I'll ask some questions via code comments.

File: src/compiler/crystal/syntax/parser.cr [hunk 1]
```diff
@@ -3702,14 +3704,19 @@ module Crystal
...
       doc = @token.doc
-
       next_token_skip_space
-      exp = parse_op_assign
 
-      modifier = VisibilityModifier.new(modifier, exp)
-      modifier.doc = doc
-      exp.doc = doc
-      modifier
+      if @token.type == :NEWLINE
```
Comment (jessedoyle) [inline src/compiler/crystal/syntax/parser.cr]: @asterite - A few questions:

1. To produce a `VisibilityModifier` with a nil expression, do you mean something like `VisibilityModifier.new(modifier, Nop.new)`?

2. To move the logic to the `TypeVisitor` (which I'm assuming is in the `visit(node : VisibilityModifier)` method in `type_inference.cr`), how would I access the subsequent `Def` nodes if the expression is a `Nop`?

Comment (asterite) [inline src/compiler/crystal/syntax/parser.cr]: 1. Maybe `VisibilityModifier.new(modifier, nil)`. That, of course, will "break" some code, but then you have: if the `exp` is `nil` it's just `private` (or `protected`), if it's not `nil` then it's applying to some other node.
2. You'll have to store the current visibility in the TypeVisitor. Start with `nil`. Then when you encounter a `VisibilityModifier` with a `nil` exp, you set the current visibility to that one. Subsequent defs or declarations (like calls, that expand to macros) should check the current visibility and apply it.

**Comment by asterite** [comment]
What I meant with what I said above (I was on a cellphone, I wasn't going to write code) is this:

```crystal
record Point, x, y do
  def foo
    1
  end

  private

  def bar
    2
  end
end
```

Since currently "private" doesn't produce an AST node, when all that code is "pasted" into the `record` macro, there's no trace of it, so `bar` ends up with public visibility.

**Comment by asterite** [comment]
Finally, the formatter I'm working on needs to know about that lonely "private" there when traversing the AST nodes.

**Comment by asterite** [comment]
After talking with @waj, we decided not to do this, mostly because of the issue of having to scroll up to check a method's visibility.

**Comment by ozra** [comment]
How about at least allowing to put visibility modifier on its own line (just before def), just a lexing/parsing adjustment, still same AST structure?
```crystal
private
def my_private_def
end
```

I like the identifying part of a method to be aligned in the same column.

**Comment by jessedoyle** [comment]
@asterite - I'll close this PR if it's never going to make it into master.

But first I'd like to make a few final arguments (with the hope that you will reconsider :smiley:):

1. The ruby-style of method visibility encourages developers to place their protected/private method definitions at the bottom of a class or module. I consider this to be a superior style with respect to private methods being defined all over a class.

2. Every time I'm forced to specify method visibility for individual methods, it feels like I'm battling the compiler. Maybe this is just me, but it just seems tedious when I'm aware of the syntax for method visibility in Ruby. It appears that other people in this thread may feel similarly, and you even said that you missed the Ruby syntax.

3. The current visibility syntax (at least to me) sets a precedence for repeated code. I believe that the syntax should discourage any code duplication, and the Ruby syntax seems to solve the issue. Also, I generally try to adhere to the 80 character limit per line, and forcing a modifier in front of every private/protected method definition makes this difficult to follow.

I hope this post doesn't come off as insincere, I just want to put a few arguments out there!

**Comment by ozra** [comment]
I for one agree on all accounts. Open an RFC in issues for a while before closing, please?

**Comment by jessedoyle** [comment]
@asterite - Thanks for the help! My last commit moved the logic from the parser over to the `TypeVisitor`. 

Even if this patch doesn't make it into Crystal it was a great chance to learn about the compiler's internals!

**Comment by asterite** [comment]
@jessedoyle Yes, I didn't close this issue because it's something to discuss and we aren't super strong about either of the solutions.

Our main worry are things like this:

```crystal
class Foo
  
  private

  def foo1
  end

  public def foo2
  end

  def foo3
  end
end
```

If you look at `foo3` it's not immediately obvious what's its visibility. Same goes with making one def private and letting the next ones inherit the current visibility, it can be confusing. But I agree about the duplication (although you could say that writing `def` over and over again is also a duplication).

Also:

> The ruby-style of method visibility encourages developers to place their protected/private method definitions at the bottom of a class or module

We don't think that's necessarily good. It's something good to keep related method together. But, of course, if we allow both styles it will be possible.

We need to review this a bit more with @waj and come to a decision.

**Comment by jessedoyle** [comment]
Sounds good, please let me know if there is anything to be done on my end.

**Comment by jessedoyle** [comment]
Sorry, I think I may have botched that last push somewhat - let me know if it causes any issues.

@asterite - I definitely understand your concern when mixing both syntaxes. I still think this feature worthwhile to have in the compiler.

We could solve any ambiguity a few ways:
- Make it clear in code style documentation that mixing and matching the two visibility syntaxes is not considered a best practice.
- Update the PR ensure that the compiler throws an error if the two syntaxes are mixed. Therefore both styles may still be used, but they must be used independently of one another.

**Comment by ozra** [comment]
I like the idea of sticking with one or the other for one type definition.

**Comment by jessedoyle** [comment]
@asterite: Have you had any change of thought on this PR?

**Comment by asterite** [comment]
@jessedoyle sorry for not replying earlier, I think I saw this but was busy and then forgot about it.

I'd like to keep with the current way a bit more time and see what happens. When I code in Ruby I don't like much having to scroll a bit to see what visibility a method has, and it still happens to me when I browse Ruby code, so I'd like to avoid that in Crystal.

**Comment by jessedoyle** [comment]
@asterite - No worries! If you decide to change your mind, I'll look into it!

**Pull request closed without merging**